README.md
<img src="http://i.imgur.com/qRyNMx4.png" width="550">

[![Build Status](https://travis-ci.org/speakeasyjs/speakeasy.svg?branch=v2)](https://travis-ci.org/speakeasyjs/speakeasy)
[![NPM downloads](https://img.shields.io/npm/dt/speakeasy.svg)](https://www.npmjs.com/package/speakeasy)
[![NPM version](https://img.shields.io/npm/v/speakeasy.svg)](https://www.npmjs.com/package/speakeasy)

---

Speakeasy is a one-time passcode generator, suitable for use in two-factor
authentication, that supports Google Authenticator and other two-factor apps.
It includes robust support for custom token lengths, authentication windows,
and other features, and includes helpers like a secret key generator.

Speakeasy implements one-time passcode generators as standardized by the
[Initiative for Open Authentication (OATH)][oath]. The HMAC-Based One-Time
Password (HOTP) algorithm defined by [RFC 4226][rfc4226] and the Time-Based
One-time Password (TOTP) algorithm defined in [RFC 6238][rfc6238] are
supported.

## Install

```sh
npm install --save speakeasy
```

## Demo

This demo uses the `generate_key` method of Speakeasy to generate a secret key,
displays a Google Authenticator–compatible QR code which you can scan into your
phone's two-factor app, and shows the token, which you can verify with your
phone. Includes sample code.

<a href="https://sedemo-mktb.rhcloud.com/"><img src="http://i.imgur.com/HN11GTW.png" width="191"></a>

## Usage

```js
var speakeasy = require("speakeasy");
var token = speakeasy.hotp({
  secret: "xyzzy",
  counter: 123
});
// token = "378764"

var ok = speakeasy.hotp.verify({
  secret: "xyzzy",
  token: token,
  counter: 123
});
// ok = {delta: 0}
```

## Documentation

Full documentation at http://speakeasyjs.github.io/speakeasy/

## License

This project incorporates code from [passcode][], which was originally a
fork of speakeasy, and [notp][], both of which are licensed under MIT.
Please see the [LICENSE](LICENSE) file for the full combined license.

Icon created by Gregor Črešnar from the Noun Project.

[passcode]: http://github.com/mikepb/passcode
[notp]: https://github.com/guyht/notp
[oath]: http://www.openauthentication.org/
[rfc4226]: https://tools.ietf.org/html/rfc4226
[rfc6238]: https://tools.ietf.org/html/rfc6238


History.md
1.0.2 / 2015-07-13
==================

  * [Fixed] Don't repeat the secret key generating the digest.


1.0.1 / 2015-07-13
==================

  * [Fixed] Ignore case on algorithm option.


1.0.0 / 2015-07-12
==================

  * Initial release based on speakeasy and notp.


test/url_test.js
"use scrict";

var assert = require('assert');
var speakeasy = require('..');
var url = require("url");

describe("#url", function () {

  it("should require options", function () {
    assert.throws(function () {
      speakeasy.url();
    });
  });

  it("should validate type", function () {
    assert.throws(function () {
      speakeasy.url({
        type: "haha",
        secret: "hello",
        label: "that",
      }, /invalid type `haha`/);
    });
  });

  it("should require secret", function () {
    assert.throws(function () {
      speakeasy.url({
        label: "that"
      }, /missing secret/);
    });
  });

  it("should require label", function () {
    assert.throws(function () {
      speakeasy.url({
        secret: "hello"
      }, /missing label/);
    });
  });

  it("should require counter for HOTP", function () {
    assert.throws(function () {
      speakeasy.url({
        type: "hotp",
        secret: "hello",
        label: "that"
      }, /missing counter/);
    });
    assert.ok(speakeasy.url({
      type: "hotp",
      secret: "hello",
      label: "that",
      counter: 0
    }));
    assert.ok(speakeasy.url({
      type: "hotp",
      secret: "hello",
      label: "that",
      counter: 199
    }));
  });

  it("should validate algorithm", function () {
    assert.throws(function () {
      speakeasy.url({
        secret: "hello",
        label: "that",
        algorithm: "hello"
      }, /invalid algorithm `hello`/);
    });
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      algorithm: "sha1"
    }));
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      algorithm: "sha256"
    }));
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      algorithm: "sha512"
    }));
  });

  it("should validate digits", function () {
    assert.throws(function () {
      speakeasy.url({
        secret: "hello",
        label: "that",
        digits: "hello"
      }, /invalid digits `hello`/);
    });
    assert.throws(function () {
      speakeasy.url({
        secret: "hello",
        label: "that",
        digits: 12
      }, /invalid digits `12`/);
    });
    assert.throws(function () {
      speakeasy.url({
        secret: "hello",
        label: "that",
        digits: "7"
      }, /invalid digits `7`/);
    });
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      digits: 6
    }));
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      digits: 8
    }));
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      digits: "6"
    }));
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      digits: "8"
    }));
  });

  it("should validate period", function () {
    assert.throws(function () {
      speakeasy.url({
        secret: "hello",
        label: "that",
        period: "hello"
      }, /invalid period `hello`/);
    });
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      period: 60
    }));
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      period: 121
    }));
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      period: "60"
    }));
    assert.ok(speakeasy.url({
      secret: "hello",
      label: "that",
      period: "121"
    }));
  });

  it("should generate an URL compatible with the Google Authenticator app", function () {
    var answer = speakeasy.url({
      secret: "JBSWY3DPEHPK3PXP",
      label: "Example:alice@google.com",
      issuer: "Example",
      encoding: "base32"
    });
    var expect = "otpauth://totp/Example:alice@google.com?secret=JBSWY3DPEHPK3PXP&issuer=Example";
    assert.deepEqual(
      url.parse(answer),
      url.parse(expect)
    );
  });

});


test/hotp_test.js
"use scrict";

var assert = require('assert');
var speakeasy = require('..');

// These tests use the information from RFC 4226's Appendix D: Test Values.
// http://tools.ietf.org/html/rfc4226#appendix-D

describe('HOTP Counter-Based Algorithm Test', function () {

  describe('normal operation with secret = \'12345678901234567890\' at counter 3', function () {
    it('should return correct one-time password', function() {
      var topic = speakeasy.hotp({secret: '12345678901234567890', counter: 3});
      assert.equal(topic, '969429');
    });
  });

  describe('another counter normal operation with secret = \'12345678901234567890\' at counter 7', function () {
    it('should return correct one-time password', function() {
      var topic = speakeasy.hotp({secret: '12345678901234567890', counter: 7});
      assert.equal(topic, '162583');
    });
  });

  describe('digits override with secret = \'12345678901234567890\' at counter 4 and digits = 8', function () {
    it('should return correct one-time password', function() {
      var topic = speakeasy.hotp({secret: '12345678901234567890', counter: 4, digits: 8});
      assert.equal(topic, '40338314');
    });
  });

  describe('hexadecimal encoding with secret = \'3132333435363738393031323334353637383930\' as hexadecimal at counter 4', function () {
    it('should return correct one-time password', function() {
      var topic = speakeasy.hotp({secret: '3132333435363738393031323334353637383930', encoding: 'hex', counter: 4});
      assert.equal(topic, '338314');
    });
  });

  describe('base32 encoding with secret = \'GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ\' as base32 at counter 4', function () {
    it('should return correct one-time password', function() {
      var topic = speakeasy.hotp({secret: 'GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQ', encoding: 'base32', counter: 4});
      assert.equal(topic, '338314');
    });
  });

  describe('base32 encoding with secret = \'12345678901234567890\' at counter 3', function () {
    it('should return correct one-time password', function() {
      var topic = speakeasy.hotp({secret: '12345678901234567890', counter: 3});
      assert.equal(topic, '969429');
    });
  });

  describe('base32 encoding with secret = \'GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZA\' as base32 at counter 1, digits = 8 and algorithm as \'sha256\'', function () {
    it('should return correct one-time password', function() {
      var topic = speakeasy.hotp({secret: 'GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZA', encoding: 'base32', counter: 1, digits: 8, algorithm: 'sha256'});
      assert.equal(topic, '46119246');
    });
  });

  describe('base32 encoding with secret = \'GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNA\' as base32 at counter 1, digits = 8 and algorithm as \'sha512\'', function () {
    it('should return correct one-time password', function() {
      var topic = speakeasy.hotp({secret: 'GEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNBVGY3TQOJQGEZDGNA', encoding: 'base32', counter: 1, digits: 8, algorithm: 'sha512'});
      assert.equal(topic, '90693936');
    });
  });

});


lib/speakeasy.js
// # speakeasy
// ### HMAC One-Time Password module for Node.js, supporting counter-based and time-based moving factors
//
// speakeasy makes it easy to implement HMAC one-time passwords, supporting both counter-based (HOTP)
// and time-based moving factors (TOTP). It's useful for implementing two-factor authentication.
// Google and Amazon use TOTP to generate codes for use with multi-factor authentication.
//
// speakeasy also supports base32 keys/secrets, by passing `base32` in the `encoding` option.
// This is useful since Google Authenticator, Google's two-factor authentication mobile app
// available for iPhone, Android, and BlackBerry, uses base32 keys.
//
// This module was written to follow the RFC memos on HTOP and TOTP:
//
// * HOTP (HMAC-Based One-Time Password Algorithm): [RFC 4226](http://tools.ietf.org/html/rfc4226)
// * TOTP (Time-Based One-Time Password Algorithm): [RFC 6238](http://tools.ietf.org/html/rfc6238)
//
// One other useful function that this module has is a key generator, which allows you to
// generate keys, get them back in their ASCII, hexadecimal, and base32 representations.
// In addition, it also can automatically generate QR codes for you, as well as the specialized
// QR code you can use to scan in the Google Authenticator mobile app.
//
// An overarching goal of this module, other than to make it very easy to implement the
// HOTP and TOTP algorithms, is to be extensively documented. Indeed, it is well-documented,
// with clear functions and parameter explanations.

var crypto = require('crypto');
var base32 = require('thirty-two');

var speakeasy = {};

// speakeasy.hotp(options)
//
// Calculates the one-time password given the key and a counter.
//
// options.key                  the key
//        .counter              moving factor
//        .length(=6)           length of the one-time password (default 6)
//        .encoding(='ascii')   key encoding (ascii, hex, or base32)
//
speakeasy.hotp = function(options) {
  // set vars
  var key = options.key;
  var counter = options.counter;
  var length = options.length || 6;
  var encoding = options.encoding || 'ascii';

  // preprocessing: convert to ascii if it's not
  if (encoding === 'hex') {
    key = speakeasy.hex_to_ascii(key);
  } else if (encoding === 'base32') {
    key = base32.decode(key);
  }

  // init hmac with the key
  var hmac = crypto.createHmac('sha1', key);

  // create an octet array from the counter
  var octet_array = new Array(8);

  var counter_temp = counter;

  for (var i = 0; i < 8; i++) {
    var i_from_right = 7 - i;

    // mask 255 over number to get last 8
    octet_array[i_from_right] = counter_temp & 255;

    // shift 8 and get ready to loop over the next batch of 8
    counter_temp = counter_temp >> 8;
  }

  // create a buffer from the octet array
  var counter_buffer = new Buffer(octet_array);

  // update hmac with the counter
  hmac.update(counter_buffer);

  // get the digest in hex format
  var digest = hmac.digest('hex');

  // convert the result to an array of bytes
  var digest_bytes = speakeasy.hexToBytes(digest);

  // compute HOTP
  // get offset
  var offset = digest_bytes[19] & 0xf;

  // calculate bin_code (RFC4226 5.4)
  var bin_code = (digest_bytes[offset] & 0x7f)   << 24
                |(digest_bytes[offset+1] & 0xff) << 16
                |(digest_bytes[offset+2] & 0xff) << 8
                |(digest_bytes[offset+3] & 0xff);

  var code = speakeasy.bin_to_string(bin_code, length);

  return(code);
};

// speakeasy.totp(options)
//
// Calculates the one-time password given the key, based on the current time
// with a 30 second step (step being the number of seconds between passwords).
//
// options.key                  the key
//        .length(=6)           length of the one-time password (default 6)
//        .encoding(='ascii')   key encoding (ascii, hex, or base32)
//        .step(=30)            override the step in seconds
//        .time                 (optional) override the time to calculate with
//        .initial_time         (optional) override the initial time
//
speakeasy.totp = function(options) {
  // set vars
  var key = options.key;
  var length = options.length || 6;
  var encoding = options.encoding || 'ascii';
  var step = options.step || 30;
  var initial_time = options.initial_time || 0; // unix epoch by default

  // get current time in seconds since unix epoch
  var time = parseInt(Date.now()/1000);

  // are we forcing a specific time?
  if (options.time) {
    // override the time
    time = options.time;
  }

  // calculate counter value
  var counter = Math.floor((time - initial_time)/ step);

  // pass to hotp
  var code = this.hotp({key: key, length: length, encoding: encoding, counter: counter});

  // return the code
  return(code);
};

// speakeasy.bin_to_string(bin, length)
//
// helper function to convert a number to a string of given length.
//
speakeasy.bin_to_string = function(bin, length) {
  var bin_str = bin.toString();

  if (bin_str.length < length) {
    // pad with 0's
    var pad = '';
    var padLength = length - bin_str.length;
    for (var j = 0; j < padLength; ++j) {
      pad += '0';
    }

    return pad + bin_str;
  } else {
    // get the chars at position bin_code - length through length chars
    var sub_start = bin_str.length - length;
    var code = bin_str.substr(sub_start, length);

    return code;
  }
}

// speakeasy.hex_to_ascii(key)
//
// helper function to convert a hex key to ascii.
//
speakeasy.hex_to_ascii = function(str) {
  // key is a string of hex
  // convert it to an array of bytes...
  var bytes = speakeasy.hexToBytes(str);

  // bytes is now an array of bytes with character codes
  // merge this down into a string
  var ascii_string = '';

  for (var i = 0; i < bytes.length; i++) {
    ascii_string += String.fromCharCode(bytes[i]);
  }

  return ascii_string;
};

// speakeasy.ascii_to_hex(key)
//
// helper function to convert an ascii key to hex.
//
speakeasy.ascii_to_hex = function(str) {
  var hex_string = '';

  for (var i = 0; i < str.length; i++) {
    hex_string += str.charCodeAt(i).toString(16);
  }

  return hex_string;
};

// speakeasy.generate_key(options)
//
// Generates a random key with the set A-Z a-z 0-9 and symbols, of any length
// (default 32). Returns the key in ASCII, hexadecimal, and base32 format.
// Base32 format is used in Google Authenticator. Turn off symbols by setting
// symbols: false. Automatically generate links to QR codes of each encoding
// (using the Google Charts API) by setting qr_codes: true. Automatically
// generate a link to a special QR code for use with the Google Authenticator
// app, for which you can also specify a name.
//
// options.length(=32)              length of key
//        .symbols(=true)           include symbols in the key
//        .qr_codes(=false)         generate links to QR codes
//        .google_auth_qr(=false)   generate a link to a QR code to scan
//                                  with the Google Authenticator app.
//        .name                     (optional) add a name. no spaces.
//                                  for use with Google Authenticator
//
speakeasy.generate_key = function(options) {
  // options
  if(!options) options = {};
  var length = options.length || 32;
  var name = options.name || "Secret Key";
  var qr_codes = options.qr_codes || false;
  var google_auth_qr = options.google_auth_qr || false;
  var symbols = true;

  // turn off symbols only when explicity told to
  if (options.symbols !== undefined && options.symbols === false) {
    symbols = false;
  }

  // generate an ascii key
  var key = this.generate_key_ascii(length, symbols);

  // return a SecretKey with ascii, hex, and base32
  var SecretKey = {};
  SecretKey.ascii = key;
  SecretKey.hex = this.ascii_to_hex(key);
  SecretKey.base32 = base32.encode(key).toString().replace(/=/g,'');

  // generate some qr codes if requested
  if (qr_codes) {
    SecretKey.qr_code_ascii = 'https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=' + encodeURIComponent(SecretKey.ascii);
    SecretKey.qr_code_hex = 'https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=' + encodeURIComponent(SecretKey.hex);
    SecretKey.qr_code_base32 = 'https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=' + encodeURIComponent(SecretKey.base32);
  }

  // generate a QR code for use in Google Authenticator if requested
  // (Google Authenticator has a special style and requires base32)
  if (google_auth_qr) {
    // first, make sure that the name doesn't have spaces, since Google Authenticator doesn't like them
    name = name.replace(/ /g,'');
    SecretKey.google_auth_qr = 'https://chart.googleapis.com/chart?chs=166x166&chld=L|0&cht=qr&chl=otpauth://totp/' + encodeURIComponent(name) + '%3Fsecret=' + encodeURIComponent(SecretKey.base32);
  }

  return SecretKey;
};

// speakeasy.generate_key_ascii(length, symbols)
//
// Generates a random key, of length `length` (default 32).
// Also choose whether you want symbols, default false.
// speakeasy.generate_key() wraps around this.
//
speakeasy.generate_key_ascii = function(length, symbols) {
  var bytes = crypto.randomBytes(length || 32);
  var set = '0123456789ABCDEFGHIJKLMNOPQRSTUVWXTZabcdefghiklmnopqrstuvwxyz';
  if (symbols) {
    set += '!@#$%^&*()<>?/[]{},.:;';
  }

  var output = '';
  for (var i = 0, l = bytes.length; i < l; i++) {
    output += set[~~(bytes[i] / 0xFF * set.length)];
  }
  return output;
};

speakeasy.hexToBytes = function (hex) {
  var bytes = [];
  for (var i = 0, l = hex.length; i < l; i += 2) {
    bytes.push(parseInt(hex.slice(i, i + 2), 16));
  }
  return bytes;
};

// alias, not the TV show
speakeasy.counter = speakeasy.hotp;
speakeasy.time = speakeasy.totp;

module.exports = speakeasy;


index.js
"use strict";

/**
 * Module dependencies.
 */

var base32 = require("base32.js");
var crypto = require("crypto");
var url = require("url");

/**
 * Digest the one-time passcode options.
 *
 * @param {Object} options
 * @param {String} options.secret Shared secret key
 * @param {Integer} options.counter Counter value
 * @param {String} [options.encoding="ascii"] Key encoding (ascii, hex,
 *   base32, base64).
 * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
 *   sha512).
 * @return {Buffer} The one-time passcode as a buffer.
 */

exports.digest = function digest (options) {
  var i;

  // unpack options
  var key = options.secret;
  var counter = options.counter;
  var encoding = options.encoding || "ascii";
  var algorithm = (options.algorithm || "sha1").toLowerCase();

  // convert key to buffer
  if (!Buffer.isBuffer(key)) {
    key = encoding == "base32" ? base32.decode(key)
                               : new Buffer(key, encoding);
  }

  // create an buffer from the counter
  var buf = new Buffer(8);
  var tmp = counter;
  for (i = 0; i < 8; i++) {

    // mask 0xff over number to get last 8
    buf[7 - i] = tmp & 0xff;

    // shift 8 and get ready to loop over the next batch of 8
    tmp = tmp >> 8;
  }

  // init hmac with the key
  var hmac = crypto.createHmac(algorithm, key);

  // update hmac with the counter
  hmac.update(buf);

  // return the digest
  return hmac.digest();
};

/**
 * Generate a counter-based one-time passcode.
 *
 * @param {Object} options
 * @param {String} options.secret Shared secret key
 * @param {Integer} options.counter Counter value
 * @param {Buffer} [options.digest] Digest, automatically generated by default
 * @param {Integer} [options.digits=6] The number of digits for the one-time
 *   passcode.
 * @param {String} [options.encoding="ascii"] Key encoding (ascii, hex,
 *   base32, base64).
 * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
 *   sha512).
 * @return {String} The one-time passcode.
 */

exports.hotp = function hotpGenerate (options) {

  // unpack options
  var digits = options.digits || 6;

  // digest the options
  var digest = options.digest || exports.digest(options);

  // compute HOTP offset
  var offset = digest[digest.length - 1] & 0xf;

  // calculate binary code (RFC4226 5.4)
  var code = (digest[offset] & 0x7f) << 24
               | (digest[offset + 1] & 0xff) << 16
               | (digest[offset + 2] & 0xff) << 8
               | (digest[offset + 3] & 0xff);

  // left-pad code
  code = new Array(digits + 1).join("0") + code.toString(10);

  // return length number off digits
  return code.substr(-digits);
};

/**
 * Verify a counter-based One Time passcode.
 *
 * @param {Object} options
 * @param {String} options.secret Shared secret key
 * @param {String} options.token Passcode to validate
 * @param {Integer} options.counter Counter value. This should be stored by
 *   the application and must be incremented for each request.
 * @param {Integer} [options.digits=6] The number of digits for the one-time
 *   passcode.
 * @param {Integer} [options.window=50] The allowable margin for the counter.
 *   The function will check "W" codes in the future against the provided
 *   passcode, e.g. if W = 10, and C = 5, this function will check the
 *   passcode against all One Time Passcodes between 5 and 15, inclusive.
 * @param {String} [options.encoding="ascii"] Key encoding (ascii, hex,
 *   base32, base64).
 * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
 *   sha512).
 * @return {Object} On success, returns an object with the counter
 *   difference between the client and the server as the `delta` property.
 * @method hotp․verify
 * @global
 */

exports.hotp.verify = function hotpVerify (options) {
  var i;

  // shadow options
  options = Object.create(options);

  // unpack options
  var token = options.token;
  var window = options.window || 50;
  var counter = options.counter || 0;

  // loop from C to C + W
  for (i = counter; i <= counter + window; ++i) {
    options.counter = i;
    if (exports.hotp(options) == token) {
      // found a matching code, return delta
      return {delta: i - counter};
    }
  }

  // no codes have matched
};

/**
 * Calculate counter value based on given options.
 *
 * @param {Object} options
 * @param {Integer} [options.time] Time with which to calculate counter value
 * @param {Integer} [options.step=30] Time step in seconds
 * @param {Integer} [options.epoch=0] Initial time since the UNIX epoch from
 *   which to calculate the counter value. Defaults to `Date.now()`.
 * @return {Integer} The calculated counter value
 * @private
 */

exports._counter = function _counter (options) {
  var step = options.step || 30;
  var time = options.time != null ? options.time : Date.now();
  var epoch = options.epoch || 0;
  return Math.floor((time - epoch) / step / 1000);
};

/**
 * Generate a time-based one-time passcode.
 *
 * @param {Object} options
 * @param {String} options.secret Shared secret key
 * @param {Integer} [options.time] Time with which to calculate counter value
 * @param {Integer} [options.step=30] Time step in seconds
 * @param {Integer} [options.epoch=0] Initial time since the UNIX epoch from
 *   which to calculate the counter value. Defaults to `Date.now()`.
 * @param {Integer} [options.counter] Counter value, calculated by default.
 * @param {Integer} [options.digits=6] The number of digits for the one-time
 *   passcode.
 * @param {String} [options.encoding="ascii"] Key encoding (ascii, hex,
 *   base32, base64).
 * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
 *   sha512).
 * @return {String} The one-time passcode.
 */

exports.totp = function totpGenerate (options) {

  // shadow options
  options = Object.create(options);

  // calculate default counter value
  if (options.counter == null) options.counter = exports._counter(options);

  // pass to hotp
  return this.hotp(options);
};

/**
 * Verify a time-based One Time passcode.
 *
 * @param {Object} options
 * @param {String} options.secret Shared secret key
 * @param {String} options.token Passcode to validate
 * @param {Integer} [options.time] Time with which to calculate counter value
 * @param {Integer} [options.step=30] Time step in seconds
 * @param {Integer} [options.epoch=0] Initial time since the UNIX epoch from
 *   which to calculate the counter value. Defaults to `Date.now()`.
 * @param {Integer} [options.counter] Counter value, calculated by default.
 * @param {Integer} [options.digits=6] The number of digits for the one-time
 *   passcode.
 * @param {Integer} [options.window=6] The allowable margin for the counter.
 *   The function will check "W" codes in the future and the past against the
 *   provided passcode, e.g. if W = 5, and C = 1000, this function will check
 *   the passcode against all One Time Passcodes between 995 and 1005,
 *   inclusive.
 * @param {String} [options.encoding="ascii"] Key encoding (ascii, hex,
 *   base32, base64).
 * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
 *   sha512).
 * @return {Object} On success, returns an object with the time step
 *   difference between the client and the server as the `delta` property.
 * @method totp․verify
 * @global
 */

exports.totp.verify = function totpVerify (options) {

  // shadow options
  options = Object.create(options);

  // unpack options
  var window = options.window != null ? options.window : 0;

  // calculate default counter value
  if (options.counter == null) options.counter = exports._counter(options);

  // adjust for two-sided window
  options.counter -= window;
  options.window += window;

  // pass to hotp.verify
  return exports.hotp.verify(options);
};

/**
 * Generate an URL for use with the Google Authenticator app.
 *
 * Authenticator considers TOTP codes  valid for 30 seconds. Additionally,
 * the app presents 6 digits codes to the user. According to the
 * documentation, the period and number of digits are currently ignored by
 * the app.
 *
 * To generate a suitable QR Code, pass the generated URL to a QR Code
 * generator, such as the `qr-image` module.
 *
 * @param {Object} options
 * @param {String} options.secret Shared secret key
 * @param {Integer} options.label Used to identify the account with which
 *   the secret key is associated, e.g. the user's email address.
 * @param {Integer} [options.type="totp"] Either "hotp" or "totp".
 * @param {Integer} [options.counter] The initial counter value, required
 *   for HOTP.
 * @param {Integer} [options.issuer] The provider or service with which the
 *   secret key is associated.
 * @param {String} [options.algorithm="sha1"] Hash algorithm (sha1, sha256,
 *   sha512).
 * @param {Integer} [options.digits=6] The number of digits for the one-time
 *   passcode. Currently ignored by Google Authenticator.
 * @param {Integer} [options.period=30] The length of time for which a TOTP
 *   code will be valid, in seconds. Currently ignored by Google
 *   Authenticator.
 * @param {String} [options.encoding] Key encoding (ascii, hex, base32,
 *   base64). If the key is not encoded in Base-32, it will be reencoded.
 * @return {String} A URL suitable for use with the Google Authenticator.
 * @see https://github.com/google/google-authenticator/wiki/Key-Uri-Format
 */

exports.url = function (options) {

  // unpack options
  var secret = options.secret;
  var label = options.label;
  var issuer = options.issuer;
  var type = (options.type || "totp").toLowerCase();
  var counter = options.counter;
  var algorithm = options.algorithm;
  var digits = options.digits;
  var period = options.period;
  var encoding = options.encoding;

  // validate type
  switch (type) {
    case "totp":
    case "hotp":
      break;
    default:
      throw new Error("invalid type `" + type + "`");
  }

  // validate required options
  if (!secret) throw new Error("missing secret");
  if (!label) throw new Error("missing label");

  // require counter for HOTP
  if (type == "hotp" && counter == null) {
    throw new Error("missing counter value for HOTP");
  }

  // build query while validating
  var query = {secret: secret};
  if (options.issuer) query.issuer = options.issuer;

  // validate algorithm
  if (algorithm != null) {
    switch (algorithm.toUpperCase()) {
      case "SHA1":
      case "SHA256":
      case "SHA512":
        break;
      default:
        throw new Error("invalid algorithm `" + algorithm + "`");
    }
    query.algorithm = algorithm.toUpperCase();
  }

  // validate digits
  if (digits != null) {
    switch (parseInt(digits, 10)) {
      case 6:
      case 8:
        break;
      default:
        throw new Error("invalid digits `" + digits + "`");
    }
    query.digits = digits;
  }

  // validate period
  if (period != null) {
    if (~~period != period) {
      throw new Error("invalid period `" + period + "`");
    }
    query.period = period;
  }

  // convert secret to base32
  if (encoding != "base32") secret = new Buffer(secret, encoding);
  if (Buffer.isBuffer(secret)) secret = base32.encode(secret);

  // return url
  return url.format({
    protocol: "otpauth",
    slashes: true,
    hostname: type,
    pathname: label,
    query: query
  });
};
